anker/src/state.rs

```diff
@@ -90,6 +90,77 @@ pub struct HistoricalStSolPrice {
     pub st_sol_price_in_ust: MicroUst,
 }
 
+#[repr(C)]
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    BorshDeserialize,
+    BorshSerialize,
+    BorshSchema,
+    Eq,
+    PartialEq,
+    Serialize,
+)]
+pub struct HistoricalStSolPriceArray(pub [HistoricalStSolPrice; POOL_PRICE_NUM_SAMPLES]);
+
+impl HistoricalStSolPriceArray {
+    /// Create new `HistorialStSolPriceArray` with slot 0 and 1 UST in each
+    /// position of the array.
+    pub fn new() -> Self {
+        HistoricalStSolPriceArray(
+            [HistoricalStSolPrice {
+                slot: 0,
+                st_sol_price_in_ust: MicroUst(1_000_000),
+            }; 5],
+        )
+    }
+
+    /// Get last price from the array.
+    pub fn last(&self) -> HistoricalStSolPrice {
+        self.0[POOL_PRICE_NUM_SAMPLES - 1]
+    }
+
+    /// Insert `st_sol_price_in_ust` at the end of the array and rotate it.
+    pub fn insert_and_rotate(&mut self, slot: Slot, st_sol_price_in_ust: MicroUst) {
+        // Maintain the invariant that samples are sorted by ascending slot number.
+        // The sample at index 0 is the oldest, so we remove it (well, move it to the
+        // end to be overwritten), and move everything else closer to the beginning
+        // of the array. Then we overwrite the last element with the current price
+        // and slot number, and we confirmed above that that slot number is larger
+        // than the slot number of the sample before it.
+        self.0.rotate_left(1);
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].slot = slot;
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].st_sol_price_in_ust = st_sol_price_in_ust;
+        assert!(self.0[POOL_PRICE_NUM_SAMPLES - 1].slot >= self.0[POOL_PRICE_NUM_SAMPLES - 2].slot);
+    }
+
+    /// Calculate the minimum amount we are willing to pay for the `StLamports`
+    /// rewards based on the median price from the historical price information.
+    pub fn calculate_minimum_price(
+        &self,
+        rewards: StLamports,
+        sell_rewards_min_out_bps: u64,
+    ) -> Result<MicroUst, ArithmeticError> {
+        let mut sorted_arr = self.0;
+        sorted_arr.sort_by(|a, b| a.st_sol_price_in_ust.cmp(&b.st_sol_price_in_ust));
+        // Get median historical price.
+        let median_price = sorted_arr[POOL_PRICE_NUM_SAMPLES / 2];
+        let minimum_ust_per_st_sol = (median_price.st_sol_price_in_ust
+            * Rational {
+                numerator: sell_rewards_min_out_bps,
+                denominator: 10_000,
+            })?;
+        let minimum_price = (rewards
+            * Rational {
+                numerator: minimum_ust_per_st_sol.0,
+                denominator: 1_000_000_000,
+            })?;
```

Comment:
Note to self: at first I thought we might pre-multiply the two rationals and get better precision, but then there is a risk of overflow in the 64-bit multiplication, so let’s keep it like this.